process/block/shardblock_test.go

```diff
@@ -1367,6 +1263,303 @@ func TestShardProcessor_ProcessBlockWithWrongMiniBlockHeaderShouldErr(t *testing
+		EpochStartMetaHash: headerHash,
+	}
+	errGetHeader := errors.New("error getting headers")
+
+	t.Run("header is not start of epoch, should not request any header", func(t *testing.T) {
+		t.Parallel()
+
+		args := CreateMockArguments(coreComponents, dataComponents, bootstrapComponents, statusComponents)
+		args.RequestHandler = &testscommon.RequestHandlerStub{
+			RequestMetaHeaderCalled: func(hash []byte) {
+				require.Fail(t, "should not try to request meta header")
+			},
+		}
+
+		shardProc, _ := blproc.NewShardProcessor(args)
+		header := &testscommon.HeaderHandlerStub{
+			IsStartOfEpochBlockCalled: func() bool {
+				return false
+			},
+		}
+		err := shardProc.RequestEpochStartInfo(header, haveTime)
+		require.Nil(t, err)
+	})
+
+	t.Run("meta epoch greater than header epoch, should not request any header", func(t *testing.T) {
+		t.Parallel()
+
+		args := CreateMockArguments(coreComponents, dataComponents, bootstrapComponents, statusComponents)
+		args.RequestHandler = &testscommon.RequestHandlerStub{
+			RequestMetaHeaderCalled: func(hash []byte) {
+				require.Fail(t, "should not try to request meta header")
+			},
+		}
+		args.EpochStartTrigger = &testscommon.EpochStartTriggerStub{
+			MetaEpochCalled: func() uint32 {
+				return headerEpoch + 1
+			},
+		}
+
+		shardProc, _ := blproc.NewShardProcessor(args)
+
+		err := shardProc.RequestEpochStartInfo(blockStartOfEpoch, haveTime)
+		require.Nil(t, err)
+	})
+
+	t.Run("is epoch start, should not request meta header", func(t *testing.T) {
+		t.Parallel()
+
+		args := CreateMockArguments(coreComponents, dataComponents, bootstrapComponents, statusComponents)
+		args.EpochStartTrigger = &testscommon.EpochStartTriggerStub{
+			IsEpochStartCalled: func() bool {
+				return true
+			},
+		}
+		args.RequestHandler = &testscommon.RequestHandlerStub{
+			RequestMetaHeaderCalled: func(hash []byte) {
+				require.Fail(t, "should not try to request meta header")
+			},
+		}
+
+		shardProc, _ := blproc.NewShardProcessor(args)
+		err := shardProc.RequestEpochStartInfo(blockStartOfEpoch, haveTime)
+		require.Nil(t, err)
+	})
+
+	t.Run("epoch start triggered while trying to request header from pool", func(t *testing.T) {
+		t.Parallel()
+
+		args := CreateMockArguments(coreComponents, dataComponents, bootstrapComponents, statusComponents)
+
+		flag := &atomicCore.Flag{}
+		args.EpochStartTrigger = &testscommon.EpochStartTriggerStub{
+			IsEpochStartCalled: func() bool {
+				if flag.IsSet() {
+					return true
+				}
+
+				flag.SetValue(true)
+				return false
+			},
+		}
+		requestsCt := &atomicCore.Counter{}
+		args.RequestHandler = &testscommon.RequestHandlerStub{
+			RequestMetaHeaderCalled: func(hash []byte) {
+				require.Equal(t, headerHash, hash)
+				requestsCt.Increment()
+			},
+		}
+
+		shardProc, _ := blproc.NewShardProcessor(args)
+		err := shardProc.RequestEpochStartInfo(blockStartOfEpoch, haveTime)
+		require.Nil(t, err)
+		require.Equal(t, int64(1), requestsCt.Get())
```

Comment:
for the same reason, this test might fail sometimes. A time.sleep is mandatory before this statement

Reply:
Added `time.Sleep` for all tests